metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,294 @@
+             : TaxonomyBase(lca_coverage_rate, kmers_discovery_rate),
+               _anno_matrix(&anno) {
+    if (!std::filesystem::exists(tax_tree_filepath)) {
+        logger->error("Can't open taxonomic tree file {}", tax_tree_filepath);
+        exit(1);
+    }
+
+    bool require_accversion_to_taxid_map = assign_label_type(_anno_matrix->get_annotation().get_all_labels()[0]);
+
+    Timer timer;
+    if (require_accversion_to_taxid_map) {
+        logger->trace("Parsing label_taxid_map file...");
+        read_accversion_to_taxid_map(label_taxid_map_filepath, _anno_matrix);
+        logger->trace("Finished label_taxid_map file in {} sec", timer.elapsed());
+    }
+
+    timer.reset();
+    logger->trace("Parsing taxonomic tree...");
+    ChildrenList tree;
+    read_tree(tax_tree_filepath, &tree);
+    logger->trace("Finished taxonomic tree read in {} sec.", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Calculating tree statistics...");
+    std::vector<TaxId> tree_linearization;
+    dfs_statistics(root_node, tree, &tree_linearization);
+    logger->trace("Finished tree statistics calculation in {} sec.", timer.elapsed());
+
+    timer.reset();
+    logger->trace("Starting rmq preprocessing...");
+    rmq_preprocessing(tree_linearization);
+    logger->trace("Finished rmq preprocessing in {} sec.", timer.elapsed());
+}
+
+void TaxonomyClsAnno::read_tree(const std::string &tax_tree_filepath, ChildrenList *tree) {
+    std::ifstream f(tax_tree_filepath);
+    if (!f.good()) {
+        logger->error("Failed to open Taxonomic Tree file {}", tax_tree_filepath);
+        exit(1);
+    }
+
+    std::string line;
+    tsl::hopscotch_map<TaxId, TaxId> full_parents_list;
+    while (getline(f, line)) {
+        if (line == "") {
+            logger->error("The Taxonomic Tree file contains empty lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        std::vector<std::string> parts = utils::split_string(line, "\t");
+        if (parts.size() <= 2) {
+            logger->error("The Taxonomic tree filepath contains incomplete lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        uint32_t act = std::stoul(parts[0]);
+        uint32_t parent = std::stoul(parts[2]);
+        full_parents_list[act] = parent;
+        node_parent[act] = parent;
+    }
+
+    std::vector<TaxId> relevant_taxids;
+    // 'considered_relevant_taxids' is used to make sure that there are no duplications in 'relevant_taxids'.
+    tsl::hopscotch_set<TaxId> considered_relevant_taxids;
+
+    if (accversion_to_taxid_map.size()) {
+        // Store only the taxonomic nodes that exists in the annotation matrix.
+        for (const pair<std::string, TaxId> &it : accversion_to_taxid_map) {
```

Comment:
pair is not an iterator
```suggestion
        for (const std::pair<std::string, TaxId> &pair : accversion_to_taxid_map) {
```